google/cloud/aiplatform/models.py

```diff
@@ -2389,3 +2433,622 @@ def export_model(
+                underscores and dashes. International characters
+                are allowed.
+                See https://goo.gl/xmQnxf for more information
+                and examples of labels.
+            encryption_spec_key_name (Optional[str]):
+                Optional. The Cloud KMS resource identifier of the customer
+                managed encryption key used to protect the model. Has the
+                form:
+                ``projects/my-project/locations/my-region/keyRings/my-kr/cryptoKeys/my-key``.
+                The key needs to be in the same region as where the compute
+                resource is created.
+
+                If set, this Model and all sub-resources of this Model will be secured by this key.
+
+                Overrides encryption_spec_key_name set in aiplatform.init.
+            staging_bucket (str):
+                Optional. Bucket to stage local model artifacts. Overrides
+                staging_bucket set in aiplatform.init.
+        Returns:
+            model: Instantiated representation of the uploaded model resource.
+        Raises:
+            ValueError: If only `explanation_metadata` or `explanation_parameters`
+                is specified.
+        """
+        # https://cloud.google.com/vertex-ai/docs/predictions/pre-built-containers#xgboost
+        XGBOOST_SUPPORTED_VERSIONS = ["0.82", "0.90", "1.1", "1.2", "1.3", "1.4"]
+        XGBOOST_CONTAINER_IMAGE_URI_TEMPLATE = (
+            "{registry}/vertex-ai/prediction/xgboost-{cpu_or_gpu}.{version}:latest"
+        )
+
+        XGBOOST_SUPPORTED_MODEL_FILE_EXTENSIONS = [
+            ".pkl",
+            ".joblib",
+            ".bst",
+        ]
+
+        if xgboost_version is None:
+            # Using the latest version
+            xgboost_version = XGBOOST_SUPPORTED_VERSIONS[-1]
+            _LOGGER.info(f"Using the {xgboost_version} version of XGBoost.")
+
+        if xgboost_version not in XGBOOST_SUPPORTED_VERSIONS:
+            _LOGGER.error(
+                f"XGBoost version {xgboost_version} is not supported. "
+                f"Supported versions: {XGBOOST_SUPPORTED_VERSIONS}"
+            )
+
+        model_file_path_obj = pathlib.Path(model_file_path)
+        if not model_file_path_obj.is_file():
+            raise ValueError(
+                f"model_file_path path must point to a file: '{model_file_path}'"
+            )
+
+        model_file_extension = model_file_path_obj.suffix
+        if model_file_extension not in XGBOOST_SUPPORTED_MODEL_FILE_EXTENSIONS:
+            _LOGGER.warning(
+                f"Only the following XGBoost model file extensions are currently supported: '{XGBOOST_SUPPORTED_MODEL_FILE_EXTENSIONS}'"
+            )
+            _LOGGER.warning(
+                "Treating the model file as a binary serialized XGBoost Booster."
+            )
+            model_file_extension = ".bst"
+
+        # Preparing model directory
+        prepared_model_dir = tempfile.mkdtemp()
```

Comment:
Done

Reply:
I've added automatic cleanup, but then reverted it.
After adding the cleanup, I've noticed that when the methods are called asynchronously, the data is often cleaned up before it is read. So I've reverted the change.
Fortunately, the OS regularly cleans the temporary files.

Reply:
That makes sense as the upload call unblocks and returns immediately. We can still support this flow by changing the following:

1. Change to a `classmethod`. That seems more appropriate instead of indirectly referencing the `Model` class through the `aiplatform` Module below.

2. Make this method optionally synchronous by decorating with `base.optional_sync` and synchronously calling `cls.upload`.

I tested these changes below:

```Python
@classmethod
@base.optional_sync()
def upload_scikit_learn_model_file(
    cls,
    ...,
    sync=True,
) -> "Model":
  ...
  with tempfile.TemporaryDirectory() as tmpdirname:
      prepared_model_file_path = pathlib.Path(tmpdirname) / (
          "model" + model_file_extension
      )
      shutil.copy(model_file_path_obj, prepared_model_file_path)

      return cls.upload(..., sync=True)
```

Hard preference to clean up the temp dir as it's the guidance in the [pydocs](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp). 

Another option is to remove the asynchronous support for now and follow up.

Reply:
Done.